pkg/storage/unified/resource/client.go

```diff
@@ -31,24 +33,138 @@ func NewResourceClient(channel *grpc.ClientConn) ResourceClient {
 	}
 }
 
-func NewLocalResourceClient(server ResourceServer) ResourceClient {
-	channel := &inprocgrpc.Channel{}
+func NewLocalResourceClient(server ResourceServer) ResourceStoreClient {
+	return &localResourceClient{server}
+}
 
-	auth := &grpcUtils.Authenticator{}
+var _ ResourceClient = &localResourceClient{}
 
-	channel.RegisterService(
-		grpchan.InterceptServer(
-			&ResourceStore_ServiceDesc,
-			grpcAuth.UnaryServerInterceptor(auth.Authenticate),
-			grpcAuth.StreamServerInterceptor(auth.Authenticate),
-		),
-		server, // Implements all the things
-	)
+type localResourceClient struct {
+	server ResourceServer
+}
 
-	cc := grpchan.InterceptClientConn(channel, grpcUtils.UnaryClientInterceptor, grpcUtils.StreamClientInterceptor)
-	return &resourceClient{
-		ResourceStoreClient: NewResourceStoreClient(cc),
-		ResourceIndexClient: NewResourceIndexClient(cc),
-		DiagnosticsClient:   NewDiagnosticsClient(cc),
+// Create implements ResourceClient.
+func (c *localResourceClient) Create(ctx context.Context, in *CreateRequest, opts ...grpc.CallOption) (*CreateResponse, error) {
+	return c.server.Create(ctx, in)
```

Comment:
The only differences I can think of are good -- like having the user and tracing headers directly rather than needing to recreate them.

The part the is clearly different with this approach is that if you try sending things in `grpc.CallOption` or use any of the grpc stream functions for Watch -- they would be ignored.  We don't have any need to use them, so I don't think it is a problem but it is the key behavior difference

Reply:
Ideally we'd want the context behavior to mirror what happens when we are going over grpc, otherwise we risk introducing subtle bugs that manifest themselves when running over grpc but not when in-process (i.e. if we rely on side effects from the context).

Reply:
Btw, we have [this PR](https://github.com/grafana/grafana/pull/93124) to introduce the new gRPC interceptors for in-proc mode ready (other than checking Ryan's report about failures when `kubernetesAggregator=true`).

Gabriel and I talked about this yesterday, and we're also in favor of using gRPC across all modes of operation, for the same reasons you stated, Dan.
That said, it is a slight inefficiency to always decode the token to reconstruct the claims (`AuthInfo`) and Ryan's approach removes it. Also, worth mentioning, we've introduced the [unsafe variant of the token verifier](https://github.com/grafana/authlib/blob/main/authn/unsafe_verifier.go) in authlib for this particular situation, and it does not check the signature which reduces a bit the computational needs.

Reply:
I'm really torn here -- something nice about always grpc, but also lots of unnecessary extra processing when this is actually running as a simple standalone service.  

I changed out test integration tests so that it can run real gRPC (not the inproc flavor) -- so i think that is better regardless how we move forward here

Reply:
I looked into why the in-proc gRPC implementation is breaking `kubernetesAggregator`. I've outlined my findings in [this comment](https://github.com/grafana/grafana/pull/93124#issuecomment-2348613327).

I think this is a good example of code divergence, which can lead to subtle bugs over time and increase the maintenance burden. We'd diverge even more if we'd choose the implementation proposed in this PR.

On the performance benefits, I'm not yet convinced that maintaining a completely different `in-proc` implementation is worth it. Here are some thoughts on this topic:
- we don't know yet if the performance of the `in-proc` gRPC implementation will be an issue.
- if using gRPC + ID tokens for `in-proc` is too slow, we'll have the same problem on-prem and in cloud.
- `IDService.SignIdentity()` is caching, thus it avoids repeating the signing operation unnecessarily.
- `in-proc` implementation does not currently check the token signature, which means it's already slightly faster than `on-prem` and `cloud`. This is so not because of the performance benefits, but rather due to how complicated it is to make `signingkeys.Service` available where we need it.

All things considered, I think having a unified approach here is ultimately more maintainable. Plus, the `in-proc` gRPC implementation will act as a "canary in the coal mine" for performance issues.

Reply:
I agree, while it's tempting to cut out the encode/decode loop and go directly to the server we're going to be better off long term keeping things consistent and avoiding subtle differences in behavior between the 2 modes as much as possible. If we find that performance isn't good enough on low-power backends we can certainly revisit and look at an approach like this again, potentially with some more guardrails to prevent things like propagation of non-grpc contents in the context or client-side side effects of server code mutating request variables.